src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -167,21 +167,25 @@ export function run(
     }
   })();
 
-  // Submit all the batches, then check the results.
+  // Submit all the batches
+  t.device.pushErrorScope('validation');
+
   const checkResults: Array<() => void> = [];
   for (let i = 0; i < cases.length; i += casesPerBatch) {
     const batchCases = cases.slice(i, Math.min(i + casesPerBatch, cases.length));
-    const checkResult = submitBatch(
-      t,
-      expressionBuilder,
-      parameterTypes,
-      returnType,
-      batchCases,
-      cfg.inputSource
+    checkResults.push(
+      submitBatch(t, expressionBuilder, parameterTypes, returnType, batchCases, cfg.inputSource)
     );
-    checkResults.push(checkResult);
   }
 
+  // Check GPU validation (shader compilation, pipeline creation, etc) before checking the results
+  const error = await t.device.popErrorScope();
```

Comment:
This looks like it will work but the code is a little fragile: since subcases run concurrently, it's critical that there are no `await`s between pushErrorScope and popErrorScope. There aren't, but since the function is async, it would be easy to accidentally break later.

We can actually make it non-async by using what we call "eventualAsyncExpectation" which allows the test to continue doing stuff asynchronously after the end of the test function. I'll open a version that does that but also make a small suggestion to this PR that would also avoid that problem.

Reply:
https://github.com/gpuweb/cts/pull/1784

Reply:
And pushed a commit to this PR which does the small change.

Reply:
Looks great. This is exactly what I was going to do after seeing your other change. Thanks.